src/dbnode/storage/index.go

```diff
@@ -1554,52 +1605,107 @@ func (i *nsIndex) queryWithSpan(
 
 	var (
 		// State contains concurrent mutable state for async execution below.
-		state = asyncQueryExecState{
-			exhaustive: true,
-		}
-		wg            sync.WaitGroup
-		totalWaitTime time.Duration
+		state   = asyncQueryExecState{}
+		wg      sync.WaitGroup
+		permits = i.permitsManager.NewPermits(ctx)
 	)
 
-	for _, block := range blocks {
-		// Capture block for async query execution below.
-		block := block
-
-		// We're looping through all the blocks that we need to query and kicking
-		// off parallel queries which are bounded by the queryWorkersPool's maximum
-		// concurrency. This means that it's possible at this point that we've
-		// completed querying one or more blocks and already exhausted the maximum
-		// number of results that we're allowed to return. If thats the case, there
-		// is no value in kicking off more parallel queries, so we break out of
-		// the loop.
-		seriesCount := results.Size()
-		docsCount := results.TotalDocsCount()
-		alreadyExceededLimit := opts.SeriesLimitExceeded(seriesCount) || opts.DocsLimitExceeded(docsCount)
-		if alreadyExceededLimit {
-			state.Lock()
-			state.exhaustive = false
-			state.Unlock()
-			// Break out if already not exhaustive.
+	blockIters := make([]*blockIter, len(blocks))
+	for b, block := range blocks {
+		iter, err := newBlockIterFn(ctx, block, query, results)
+		if err != nil {
+			return false, err
+		}
+		blockIters[b] = &blockIter{
+			iter:       iter,
+			iterCloser: x.NewSafeCloser(iter),
+			block:      block,
+		}
+	}
+
+	defer func() {
+		for _, iter := range blockIters {
+			// safe to call Close multiple times, so it's fine to eagerly close in the loop below and here.
+			_ = iter.iterCloser.Close()
+		}
+	}()
+
+	// waitForPermit waits for a permit. returns true if the permit was acquired and the wait time.
+	waitForPermit := func() (bool, time.Duration) {
+		queryCanceled := func() bool {
+			return opts.LimitsExceeded(results.Size(), results.TotalDocsCount()) || state.hasErr()
+		}
```

Comment:
Can this be defined above `waitForPermit`? That way it closure need to be allocated each time waitForPermit is called.

Also could we just add `opts` and `results` to the `state` so that state can implement this as a method? i.e. `state.cancelled()`? That way the closure doesn't have to be allocated separately and the allocation for the `state` happens just once.